### PR TITLE
[REF-6057] mcenter-cli: using cliff cli framework

### PR DESCRIPTION
**Pull request by lior-amar-pm**
Testing done:

Manual testing of the command line options,

### CODE REVIEW

File: mcenter_cli/parallelm/mcenter_cli/command_action_log.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+from parallelm.mcenter_cli.list_actions import list_actions
+
+
+class ActionLogListCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(ActionLogListCommand, self).get_parser(prog_name)
+        parser.add_argument('--object', action='store', help='List of Objects for which to get actions')
+        parser.add_argument('--action', action='store', help='List of actions to query')
+        parser.add_argument('--username', action='store', help='List of User, whose actions should be queried')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_action_log.py]: "List of User" doesn't make sense to me. The other two do make sense -- List of objections and list of actions. Should this be "List of Users"?

File: mcenter_cli/parallelm/mcenter_cli/command_action_log.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+from cliff.command import Command
+
+from parallelm.mcenter_cli.list_actions import list_actions
+
+
+class ActionLogListCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(ActionLogListCommand, self).get_parser(prog_name)
+        parser.add_argument('--object', action='store', help='List of Objects for which to get actions')
+        parser.add_argument('--action', action='store', help='List of actions to query')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_action_log.py]: "actions" should be "Actions" to follow the pattern. "List of Objects", for example. Capitalize the word is my point, or done capitalize, but be consistent.

File: mcenter_cli/parallelm/mcenter_cli/command_agent.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+from cliff.command import Command
+
+
+class AgentRegisterCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(AgentRegisterCommand, self).get_parser(prog_name)
+        parser.add_argument('agent', action='store', help='Agent address to register')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_agent.py]: Maybe provide an example on what is expected here -- for the agent address.

File: mcenter_cli/parallelm/mcenter_cli/command_agent.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+from cliff.command import Command
+
+
+class AgentRegisterCommand(Command):
+    def get_parser(self, prog_name):
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_agent.py:5]: Shouldn't you be including Pydoc for your method defs? Example:

"""
       Blah, blah, blah...

        :param prog_name: Name of program.
           
        :return: 
        :raise:
        """

File: mcenter_cli/parallelm/mcenter_cli/command_ee.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+class EEUploadCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(EEUploadCommand, self).get_parser(prog_name)
+        parser.add_argument('eefile', action='store', help='File to use for loading the EE')
+        parser.add_argument("--name", action="store", default=None,
+                            help="Name to give the EE when uploading. Existing name of EE is overwritten")
+        parser.add_argument("--agent", action="store", default=None,
+                            help="Name of agent to use when uploading the EE. Existing agent name is overwritten")
+        parser.add_argument('--force', action='store_true',
+                            help='force loading of EE, if EE already exists remove it first')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_ee.py]: Suggestion -- change to: 'Force loading of the EE. If EE already exists, then remove it'

File: mcenter_cli/parallelm/mcenter_cli/command_ee.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    def take_action(self, parsed_args):
+        ee_upload(self.app.mclient, parsed_args)
+
+
+class EEDownloadCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(EEDownloadCommand, self).get_parser(prog_name)
+        parser.add_argument('name', action='store', help='name of EE to download')
+        parser.add_argument('eefile', action='store', help='file to use for saving the EE to')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_ee.py]: Suggestion - change to: 'File to use for saving the EE'

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+        mlapp_info = mlapp_info[0]
+        self.app.stdout(pprint.pformat(mlapp_info))
+
+
+class MLAppUploadCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(MLAppUploadCommand, self).get_parser(prog_name)
+        parser.add_argument('name', action='store', help='Name of MLApp to query for info')
+
+        parser.add_argument('appdir', action='store', help='directory to use for loading the MLApp')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_mlapp.py]: Please change the first letter in all your help strings to be caps, so as to be consistent across all your command help strings.

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+        self.app.stdout("Uploading mlapp from {}".format(parsed_args.appdir))
+        if parsed_args.mlapp_version == MLAppVersions.V2:
+            upload_mlapp_v2(mclient=self.app.mclient,
+                            mlapp_dir=parsed_args.appdir,
+                            force=parsed_args.force,
+                            ee=parsed_args.ee)
+        elif parsed_args.mlapp_version == MLAppVersions.V1:
+            upload_mlapp(self.app.mclient, parsed_args.appdir)
+
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_mlapp.py:71]: do we need an else: case here, just incase the parsed_args.mlapp_version is something other than V1 or V2? I know we're controlling the versions, but couldn't hurt to have that case covered. I'm a paranoid programmer. :)

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    def take_action(self, parsed_args):
+        self.app.stdout("Deleting MLApp: {}".format(parsed_args.name))
+        MLAppDeleteHelper(self.app.mclient, parsed_args.name, False).delete()
+
+
+class MLAppDownloadCommand(Command):
+    def get_parser(self, prog_name):
+        parser = super(MLAppDownloadCommand, self).get_parser(prog_name)
+        parser.add_argument('name', action='store', help='name of MLApp to download')
+        parser.add_argument('appdir', action='store', help='directory to use for saving the MLApp to')
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/command_mlapp.py]: Suggestion - change to: 'Directory to use for saving the MLApp'.

Point is on some of these, not proper to end a sentence with a preposition.

File: mcenter_cli/parallelm/mcenter_cli/mcli.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+from cliff.app import App
+from cliff.commandmanager import CommandManager
+
+from parallelm.mcenter_cli import version
+from parallelm.mcenter_client.mcenter_client import MCenterClient
+
+
+class MCli(App):
+
+    DEFAULT_LOG_LEVEL = "error"
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/mcli.py]: why is DEFAULT_LOG_LEVEL set to "error"? Thought that would be typically set "warning".

File: mcenter_cli/parallelm/mcenter_cli/mcli.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+            description='mcli - MCenter CLI',
+            version=version,
+            command_manager=CommandManager('mcenter.cli'),
+            deferred_help=True,
+        )
+
+        self._prog_name = prog_name
+        self._options = None
+        self.mclient = None
+        self._logging_level = logging.ERROR
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/mcli.py]: think this should be WARNING

File: mcenter_cli/parallelm/mcenter_cli/mcli.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+        self._commands = {}
+        self._example_text = '''Examples:
+            {prog_name} --server localhost mlapp-load /opt/mlapps/regression-prod-v32 
+            {prog_name} mlapp-download my-ml-app  /opt/mlapps/my-ml-app-new
+            {prog_name} mlapp-delete my-ml-app
+            {prog_name} get-action-log --object=MLApp,Model --action=Stop,Upload --username=admin,MLOPSUser --status=success
+            '''.format(prog_name=prog_name)
+        self._argv_0 = sys.argv[0]
+
+        # self._init_parser(logo)
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/mcli.py]: Need this commented out code -- lines 37 to 46?

File: mcenter_cli/parallelm/mcenter_cli/mcli.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+
+        logging.basicConfig(level=self._logging_level)
+
+        if not self.options.deferred_help:
+            self.mclient = MCenterClient(server=self.options.server,
+                                         user=self.options.user,
+                                         password=self.options.password)
+            res = self.mclient.login()
+            if res is None:
+                print("Unable to connect to MCenter. Please check server address, user name and password provided")
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/mcli.py]: remove "provided" from this sentence.

File: mcenter_cli/parallelm/mcenter_cli/mcli.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+        self.LOG.debug('initialize_app')
+        self._general_actions()
+
+    def prepare_to_run_command(self, cmd):
+        self.LOG.debug('prepare_to_run_command %s', cmd.__class__.__name__)
+
+    def clean_up(self, cmd, result, err):
+        self.LOG.debug('clean_up %s', cmd.__class__.__name__)
+        if err:
+            self.LOG.debug('got an error: %s', err)
```
Comment (kwatson88) [inline mcenter_cli/parallelm/mcenter_cli/mcli.py]: 'got an error' is kinda small, and may be hard to see if a bunch of output/stack trace. Suggestion -- change to something like, "***Error: %s"

**Review by kwatson88 — APPROVED** [review]

File: mcenter_cli/parallelm/mcenter_cli/command_action_log.py (hunk 1 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_action_log.py]: Fixed

File: mcenter_cli/parallelm/mcenter_cli/command_agent.py (hunk 3 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_agent.py]: Fixed

File: mcenter_cli/parallelm/mcenter_cli/command_agent.py (hunk 4 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_agent.py:5]: All the classes inheriting from the Command are basically overriding the base class methods. So the doctoring for the get_parser and take_action would be the same all over..

File: mcenter_cli/parallelm/mcenter_cli/command_ee.py (hunk 5 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_ee.py]: fixed

File: mcenter_cli/parallelm/mcenter_cli/command_ee.py (hunk 6 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_ee.py]: fixed

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py (hunk 7 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_mlapp.py]: Done

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py (hunk 8 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_mlapp.py:71]: Done.

File: mcenter_cli/parallelm/mcenter_cli/command_mlapp.py (hunk 9 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_mlapp.py]: done

File: mcenter_cli/parallelm/mcenter_cli/mcli.py (hunk 10 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/mcli.py]: Fixed

File: mcenter_cli/parallelm/mcenter_cli/mcli.py (hunk 11 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/mcli.py]: done

File: mcenter_cli/parallelm/mcenter_cli/mcli.py (hunk 12 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/mcli.py]: done

File: mcenter_cli/parallelm/mcenter_cli/mcli.py (hunk 13 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/mcli.py]: done

File: mcenter_cli/parallelm/mcenter_cli/mcli.py (hunk 14 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/mcli.py]: done

File: mcenter_cli/setup.py [hunk 15]
```diff
@@ -8,7 +8,7 @@
 desc = "MCenter command line tool"
 lic = "ParallelM"
 
-required_python_pkgs = ["requests"]
+required_python_pkgs = ["requests", "cliff"]
```
Comment (yakov-g) [inline mcenter_cli/setup.py:11]: Don't you want to increment version number?

**Review by yakov-g — APPROVED** [review]

File: mcenter_cli/parallelm/mcenter_cli/command_action_log.py (hunk 2 above)
Comment (lior-amar-pm) [reply mcenter_cli/parallelm/mcenter_cli/command_action_log.py]: fixed

**Pull request merged**